### PR TITLE
UART driver refactor

### PR DESCRIPTION
**Pull request by fauxpark**
## Description

Moves the existing `uart.[ch]` driver to `drivers/`, and adds a ChibiOS counterpart (which may need some QA from someone more versed in the ARM side of things).

It would also be good to get some `puts()` and `gets()` implementations added, if possible. And I'm not really sure what to do with `uart_available()` on ARM.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by fauxpark** [comment]
Working on AVR (tested Pro Micro), Bluepill F103, and Proton-C, though for the latter, USART2 is actually enabled by default, despite the fact that A9 and A10 (labelled TX and RX on the pinout) are AFs of USART1. Should I change this?

File: drivers/chibios/uart.c [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+uint8_t uart_getchar(void) {
+    msg_t res = sdGet(&SERIAL_DRIVER);
+
+    return (uint8_t)res;
+}
+
+uint8_t uart_available(void) {
+    // TODO
+    return 0;
```
Comment (tzarc) [inline drivers/chibios/uart.c:59]: How about:
```suggestion
    return sdGetWouldBlock(&SERIAL_DRIVER);
```

**Review by tzarc — APPROVED** [review]

File: docs/uart_driver.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
+# UART Driver
+
+The UART drivers used in QMK have a set of common functions to allow portability between MCUs.
+
+Currently, this driver does not support enabling hardware flow control (the `RTS` and `CTS` pins) if available, but will do so in future.
```
Comment (tzarc) [inline docs/uart_driver.md:5]:
```suggestion
Currently, this driver does not support enabling hardware flow control (the `RTS` and `CTS` pins) if available, but may do so in future.
```

File: docs/uart_driver.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+|ATmega16/32U4|`D3`|`D2`|`D5` |`B7` |
+|AT90USB64/128|`D3`|`D2`|*n/a*|*n/a*|
+|ATmega32A    |`D1`|`D0`|*n/a*|*n/a*|
+|ATmega328/P  |`D1`|`D0`|*n/a*|*n/a*|
+
+## ChibiOS/ARM Configuration
+
+You'll need to determine which pins can be used for UART -- as an example, STM32 parts generally have multiple UART peripherals, labeled USART1, USART2, USART3 etc.
+
+To enable UART, modify your board's `mcuconf.h` to enable the peripheral you've chosen -- in the case of using USART2, modify `STM32_SERIAL_USE_USART2` to be `TRUE`.
```
Comment (tzarc) [inline docs/uart_driver.md:23]:
```suggestion
To enable UART, modify your board's `mcuconf.h` to enable the peripheral you've chosen -- in the case of using USART2, modify `STM32_SERIAL_USE_USART2` to be `TRUE`. If not using a Proton-C, you will also need to supply a `halconf.h`, which includes the definition `#define HAL_USE_SERIAL TRUE`.
```

**Review by tzarc — CHANGES_REQUESTED** [review]

**Review by tzarc — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Pull request merged**